app/src/main/java/edu/cmu/pocketsphinx/demo/PocketSphinxActivity.java

```diff
@@ -49,70 +54,104 @@
 
 public class PocketSphinxActivity extends Activity implements
         RecognitionListener {
-		
+
     /* Named searches allow to quickly reconfigure the decoder */
     private static final String KWS_SEARCH = "wakeup";
     private static final String FORECAST_SEARCH = "forecast";
     private static final String DIGITS_SEARCH = "digits";
     private static final String PHONE_SEARCH = "phones";
     private static final String MENU_SEARCH = "menu";
-    
+
     /* Keyword we are looking for to activate menu */
     private static final String KEYPHRASE = "oh mighty computer";
 
+    private static final int MY_PERMISSIONS_REQUEST_RECORD_AUDIO = 1;
+
     private SpeechRecognizer recognizer;
     private HashMap<String, Integer> captions;
 
     @Override
     public void onCreate(Bundle state) {
         super.onCreate(state);
 
-        // Prepare the data for UI
-        captions = new HashMap<String, Integer>();
-        captions.put(KWS_SEARCH, R.string.kws_caption);
-        captions.put(MENU_SEARCH, R.string.menu_caption);
-        captions.put(DIGITS_SEARCH, R.string.digits_caption);
-        captions.put(PHONE_SEARCH, R.string.phone_caption);
-        captions.put(FORECAST_SEARCH, R.string.forecast_caption);
-        setContentView(R.layout.main);
-        ((TextView) findViewById(R.id.caption_text))
-                .setText("Preparing the recognizer");
-
-        // Recognizer initialization is a time-consuming and it involves IO,
-        // so we execute it in async task
-
-        new AsyncTask<Void, Void, Exception>() {
-            @Override
-            protected Exception doInBackground(Void... params) {
-                try {
-                    Assets assets = new Assets(PocketSphinxActivity.this);
-                    File assetDir = assets.syncAssets();
-                    setupRecognizer(assetDir);
-                } catch (IOException e) {
-                    return e;
+        // Check if user has given permission to record audio
+        int permissionCheck = ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.RECORD_AUDIO);
+        if (permissionCheck == PackageManager.PERMISSION_DENIED) {
+            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO}, MY_PERMISSIONS_REQUEST_RECORD_AUDIO);
+        } else {
```

Comment:
here you can have `return` to exit early instead of `else` to avoid extra nesting level for the rest of the code.